lib/relooper/src/graph/enrichments.rs

```diff
@@ -1,8 +1,27 @@
 use crate::graph::cfg::{Cfg, CfgEdge, CfgLabel};
 use crate::traversal::graph::bfs::Bfs;
-use crate::traversal::graph::dfs::{Dfs, DfsPost, DfsPostReverseInstantiator};
-use std::collections::{HashMap, HashSet, VecDeque};
-use std::vec::Vec;
+use std::collections::HashSet;
+
+use super::{domtree::DomTree, node_ordering::NodeOrdering, Graph};
+
+struct Lazy<T, F> {
```

Comment:
Probably worth having a comment somewhere that an invariant of the struct is that exactly one of `init` or `value` is set at all times. Maybe there is an alternative design of this data type where that invariant is enforced within the type itself (using enums), but it's probably not a big deal to leave it as is.

Reply:
changing lazy to enum would require some uninitialized memory magic which (as soon as I remember) I havent figured out
Ill leave a comment here

Reply:
You can get around using uninitialized memory by requiring `T: Default` and using `std::mem::take`. I would implement this type as

```rust
enum Lazy<T, F> {
    Resolved(T),
    Thunk(F),
}

impl<T: Default, F> Default for Lazy<T, F> {
    fn default() -> Self {
        Self::Resolved(T::default())
    }
}

impl<T: Default, F: FnOnce() -> T> Lazy<T, F> {
    pub fn force(&mut self) -> &T {
        if let Self::Resolved(value) = self {
            return value;
        }
        
        let thunk = std::mem::take(self);
        match thunk {
            Self::Thunk(f) => {
                let value = f();
                *self = Self::Resolved(value);
                match self {
                    Self::Resolved(value) => value,
                    Self::Thunk(_) => unreachable!("self re-assigned to Resolved"),
                }
            }
            Self::Resolved(_) => unreachable!("Handled by early return"),
        }
    }
}
```

But I don't know if this version has worse performance because of using `default`.

It's also probably not very important to have a carefully typed version of `Lazy` since it is only used internally anyway (we're not trying to make a library out of it).

Reply:
Ive implemented it with use of Option to get rid of Default requirement on mem::take/mem::replace operations, does that looks good to you?

Reply:
The `Option` value does feel awkward to me because it's only `None` in a "transient" state as `force` is executed. It would be nice if those implementation details were hidden from the main type somehow. But that said, it's probably not worth investing more time into this since it's a private data type anyway.